File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/UserInformation.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import java.nio.ByteBuffer;
import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.kaillera.model.KailleraUser;
import org.emulinker.kaillera.relay.KailleraRelay;
import org.emulinker.util.EmuUtil;

@AutoValue
public abstract class UserInformation extends V086Message {
  public static final byte ID = 0x03;
  private static final String DESC = "User Information";

  public abstract String username();

  public abstract String clientType();

  public abstract byte connectionType();

  public static AutoValue_UserInformation create(
      int messageNumber, String username, String clientType, byte connectionType)
      throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);

    if (connectionType < 1 || connectionType > 6) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: connectionType out of acceptable range: " + connectionType);
    }

    return new AutoValue_UserInformation(
        messageNumber, ID, DESC, username, clientType, connectionType);
  }

  @Override
  public int getBodyLength() {
    return getNumBytes(username()) + getNumBytes(clientType()) + 3;
  }

  // TODO(nue): Get rid of this.
  @Override
  public final String toString() {
    return getInfoString()
        + "[userName="
        + username()
        + " clientType="
        + clientType()
        + " connectionType="
        + KailleraUser.CONNECTION_TYPE_NAMES[connectionType()]
        + "]";
  }

  @Override
  public void writeBodyTo(ByteBuffer buffer) {
    EmuUtil.writeString(buffer, username(), 0x00, KailleraRelay.config.charset());
    EmuUtil.writeString(buffer, clientType(), 0x00, KailleraRelay.config.charset());
    buffer.put(connectionType());
  }

  public static UserInformation parse(int messageNumber, ByteBuffer buffer)
      throws ParseException, MessageFormatException {
    if (buffer.remaining() < 5) throw new ParseException("Failed byte count validation!");

    String userName = EmuUtil.readString(buffer, 0x00, KailleraRelay.config.charset());

    if (buffer.remaining() < 3) throw new ParseException("Failed byte count validation!");

    String clientType = EmuUtil.readString(buffer, 0x00, KailleraRelay.config.charset());

    if (buffer.remaining() < 1) throw new ParseException("Failed byte count validation!");

    byte connectionType = buffer.get();

    return UserInformation.create(messageNumber, userName, clientType, connectionType);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/UserInformation.kt
package org.emulinker.kaillera.controller.v086.protocol

import java.nio.ByteBuffer
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.kaillera.controller.v086.protocol.V086Message.Companion.validateMessageNumber
import org.emulinker.kaillera.model.ConnectionType
import org.emulinker.kaillera.pico.AppModule
import org.emulinker.util.EmuUtil

data class UserInformation
    @Throws(MessageFormatException::class)
    constructor(
        override val messageNumber: Int,
        val username: String,
        val clientType: String,
        val connectionType: ConnectionType
    ) : V086Message() {

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
  }

  override val bodyLength: Int
    get() = getNumBytes(username) + getNumBytes(clientType) + 3

  public override fun writeBodyTo(buffer: ByteBuffer) {
    EmuUtil.writeString(buffer, username, 0x00, AppModule.charsetDoNotUse)
    EmuUtil.writeString(buffer, clientType, 0x00, AppModule.charsetDoNotUse)
    buffer.put(connectionType.byteValue)
  }

  companion object {
    const val ID: Byte = 0x03

    @Throws(ParseException::class, MessageFormatException::class)
    fun parse(messageNumber: Int, buffer: ByteBuffer): UserInformation {
      if (buffer.remaining() < 5) throw ParseException("Failed byte count validation!")
      val userName = EmuUtil.readString(buffer, 0x00, AppModule.charsetDoNotUse)
      if (buffer.remaining() < 3) throw ParseException("Failed byte count validation!")
      val clientType = EmuUtil.readString(buffer, 0x00, AppModule.charsetDoNotUse)
      if (buffer.remaining() < 1) throw ParseException("Failed byte count validation!")
      val connectionType = buffer.get()
      return UserInformation(
          messageNumber, userName, clientType, ConnectionType.fromByteValue(connectionType))
    }
  }
}
